File: ios-app/Sources/RedCodeNetworking/APIEndpoint.swift
import Foundation
import RedCodeCore

public struct APIEndpoint: Equatable, Sendable {
    public let method: HTTPMethod
    public let path: String
    public let queryItems: [URLQueryItem]

    public init(
        method: HTTPMethod = .get,
        path: String,
        queryItems: [URLQueryItem] = []
    ) {
        self.method = method
        self.path = path
        self.queryItems = queryItems
    }

    public func url(in environment: RedCodeEnvironment) throws -> URL {
        guard !path.isEmpty else {
            throw RedCodeError.configuration("API endpoint path cannot be empty")
        }

        guard var components = URLComponents(
            url: environment.apiBaseURL,
            resolvingAgainstBaseURL: false
        ) else {
            throw RedCodeError.configuration("Invalid API base URL")
        }

        let basePath = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let endpointPath = path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        components.path = "/" + [basePath, endpointPath]
            .filter { !$0.isEmpty }
            .joined(separator: "/")
        components.queryItems = queryItems.isEmpty ? nil : queryItems

        guard let url = components.url else {
            throw RedCodeError.configuration("Unable to build API endpoint URL")
        }

        return url
    }
}

File: android-app/app/src/main/java/com/redcode/im/androidapp/network/APIEndpoint.kt
package com.redcode.im.androidapp.network

import com.redcode.im.androidapp.core.config.RedCodeEnvironment

data class APIEndpoint(
    val method: HTTPMethod,
    val path: String,
) {
    init {
        require(path.startsWith("/")) { "API path 必须以 / 开头" }
    }

    fun url(environment: RedCodeEnvironment): String =
        environment.apiBaseUrl.trimEnd('/') + path
}
